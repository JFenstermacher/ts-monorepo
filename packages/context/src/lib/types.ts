export type EnvironmentLong = "sandbox" | "staging" | "uat" | "production" | "ephemeral" | "organization";
export type EnvironmentShort = "sandbox" | "staging" | "uat" | "prod" | "org" | "ephm";
export type Environment = EnvironmentLong | EnvironmentShort;

export type StringCase = "title" | "lowercase" | "uppercase"

export type ContextIdAttributes = {
  project: string
  service: string
  environment: Environment
  organization: string
  name: string
  attributes: string[]
}

export type IdAttributeKey = keyof ContextIdAttributes
export type IdAttributeKeys = IdAttributeKey[]

export type ContextMetaAttributes = {
  labelOrder: IdAttributeKeys
  delimiter: string
  idLengthLimit: number
  tags: Tags
  labelKeyCase: StringCase
  labelValueCase: StringCase
  labelsAsTags: IdAttributeKeys
  enabled: boolean
}

export type Tags = Record<string, string>

export type ContextType = Partial<ContextIdAttributes> & ContextMetaAttributes
export type ContextInput = Partial<ContextIdAttributes> & Partial<ContextMetaAttributes>
export type ExtractableContext = ContextInput & {
  [k: string]: any
}

export type ContextKey = keyof ContextType
export type ContextKeys = ContextKey[]
