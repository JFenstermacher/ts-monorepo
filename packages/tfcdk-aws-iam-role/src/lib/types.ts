import * as aws from '@cdktf/provider-aws';
import { AwsProvider } from "@cdktf/provider-aws/lib/provider"
import { ConstructConfig } from '@JFenstermacher/tfcdk-common';

export type PrincipalTypes = "AWS" | "Service" | "Federated" | "CanonicalUser" | "*"
export type PrincipalIdentifiers = Record<keyof PrincipalTypes, string[]>

export type AssumeRoleConfig = {
  principals?: PrincipalIdentifiers
  actions?: string[]
  config?: aws.dataAwsIamPolicyDocument.DataAwsIamPolicyDocumentConfig
}

export type AwsIamRoleConfig = ConstructConfig & {
  provider?: AwsProvider
  assumeRolePrincipals?: PrincipalIdentifiers
  assumeRoleActions?: string[]
  roleConfig: aws.dataAwsIamPolicyDocument.DataAwsIamPolicyDocumentConfig
}
