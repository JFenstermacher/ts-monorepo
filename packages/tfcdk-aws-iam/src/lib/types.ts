import { AwsProvider } from "@cdktf/provider-aws/lib/provider"
import { ConstructProps } from '@JFenstermacher/tfcdk-commons';
import { DataAwsIamPolicyDocumentConfig } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';

export type Effect = "Allow" | "Deny"
export type PrincipalTypes = "AWS" | "Service" | "Federated" | "CanonicalUser" | "*"
export type Principals = Record<keyof PrincipalTypes, string[]>

export type Condition = {
  [operator: string]: {
    [key: string]: string | string[]
  }
}

export type PolicyStatement = {
  sid?: string
  actions: string[]
  effect?: Effect
  condition?: Condition[]
}

export type Policy = PolicyStatement[]

export type AssumeRolePolicyStatement = Partial<PolicyStatement> & {
  principals: Principals
}

export type AssumeRolePolicy = AssumeRolePolicyStatement[]

export type InlinePolicy = Record<string, DataAwsIamPolicyDocumentConfig>

export type AwsIamRoleProps = ConstructProps & {
  provider?: AwsProvider
  assumeRolePolicy: DataAwsIamPolicyDocumentConfig
  inlinePolicy?: InlinePolicy
  managedPolicies: string[]
  maxSessionDuration?: number
  permissionsBoundary?: string
}
