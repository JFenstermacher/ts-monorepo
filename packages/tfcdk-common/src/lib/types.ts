import { ContextInput, ContextType } from '@JFenstermacher/context';

export type BaseConstructConfig = ContextInput & {
  context?: ContextType
}

export type ConstructConfig = BaseConstructConfig & Record<string, any>
