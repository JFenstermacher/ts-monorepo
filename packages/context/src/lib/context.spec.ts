import Context from './context';
import { ContextInput } from './types';

describe('Context', () => {
  const baseContext = new Context({
    project: "project",
    service: "service",
    environment: "sandbox"
  });

  const getContext = (contextInput: Partial<ContextInput>) => new Context({
    ...baseContext.context,
    ...contextInput
  })

  it('should be able to dump context with defualts merged', () => {
    expect(baseContext.context).toMatchSnapshot();
  })

  it('should fail validation if labelOrder has no values', () => {
    expect(() => getContext({
      labelOrder: []
    })).toThrowError();
  })

  it('should write labels as tags', () => {
    expect(baseContext.tags).toMatchSnapshot();
  })

  it('should filter tags based on labelsAsTags', () => {
    const context = getContext({
      labelsAsTags: ["name"]
    })

    expect(context.tags).toMatchSnapshot();
  })

  it('should merge user tags over default tags', () => {
    const context = getContext({
      tags: {
        service: "huh",
        environment: "what"
      }
    })

    expect(context.tags).toMatchSnapshot();
  })

  it('should allow changing tag value case', () => {
    const context = getContext({
      labelValueCase: "uppercase",
      attributes: ["attr1", "attr2"],
      labelsAsTags: [...baseContext.labelsAsTags, "attributes"]
    })

    expect(context.tags).toMatchSnapshot();
  })
});
