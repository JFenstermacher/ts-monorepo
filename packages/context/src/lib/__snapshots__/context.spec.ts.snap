// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context should allow changing tag value case 1`] = `
Object {
  "Attributes": "ATTR1,ATTR2",
  "Environment": "SANDBOX",
  "Name": "SERVICE-SANDBOX",
  "Project": "PROJECT",
  "Service": "SERVICE",
}
`;

exports[`Context should be able to dump context with defualts merged 1`] = `
Object {
  "attributes": undefined,
  "delimiter": "-",
  "enabled": true,
  "environment": "sandbox",
  "idLengthLimit": 255,
  "labelKeyCase": "title",
  "labelOrder": Array [
    "service",
    "environment",
    "name",
  ],
  "labelValueCase": "lowercase",
  "labelsAsTags": Array [
    "project",
    "service",
    "environment",
    "name",
  ],
  "name": undefined,
  "organization": "JFenstermacher",
  "project": "project",
  "service": "service",
  "tags": Object {},
}
`;

exports[`Context should filter tags based on labelsAsTags 1`] = `
Object {
  "Name": "service-sandbox",
}
`;

exports[`Context should merge user tags over default tags 1`] = `
Object {
  "Environment": "what",
  "Name": "service-sandbox",
  "Project": "project",
  "Service": "huh",
}
`;

exports[`Context should write labels as tags 1`] = `
Object {
  "Environment": "sandbox",
  "Name": "service-sandbox",
  "Project": "project",
  "Service": "service",
}
`;
